esp-lp-hal/src/i2c.rs

```diff
@@ -0,0 +1,587 @@
+                self.add_cmd_lp(
+                    &mut cmd_iterator,
+                    Command::Read {
+                        ack_value: Ack::Ack,
+                        length: (fifo_size - 1) as u8,
+                    },
+                )?;
+                // Read the last byte and send NACK
+                self.add_cmd_lp(
+                    &mut cmd_iterator,
+                    Command::Read {
+                        ack_value: Ack::Nack,
+                        length: 1,
+                    },
+                )?;
+                // Send STOP command after reading
+                self.add_cmd_lp(&mut cmd_iterator, Command::Stop)?;
+            } else {
+                // This means we have to read data more than we can fit into the Rx FIFO
+                // Read fifo_size bytes and send ACKs
+                self.add_cmd_lp(
+                    &mut cmd_iterator,
+                    Command::Read {
+                        ack_value: Ack::Ack,
+                        length: fifo_size as u8,
+                    },
+                )?;
+                // Send END command signaling more data to come
+                self.add_cmd_lp(&mut cmd_iterator, Command::End)?;
+            }
+
+            self.lp_i2c_update();
+
+            // Initiate I2C transfer
+            self.i2c.ctr().modify(|_, w| w.trans_start().set_bit());
+
+            // Await for completion (This function or mechanism should handle waiting for
+            // the I2C transfer to complete)
+            self.wait_for_completion()?;
+
+            // Read from FIFO into the current chunk
+            for byte in buffer.iter_mut() {
+                *byte = self.read_fifo();
+            }
+        }
+        Ok(())
+    }
+
+    fn master_write_read(
+        &mut self,
+        addr: u8,
+        bytes: &[u8],
+        buffer: &mut [u8],
+    ) -> Result<(), Error> {
+        // it would be possible to combine the write and read
+        // in one transaction but filling the tx fifo with
+        // the current code is somewhat slow even in release mode
+        // which can cause issues
+        self.master_write(addr, bytes)?;
+        self.master_read(addr, buffer)?;
+        Ok(())
+    }
+
+    /// Update I2C configuration
+    fn lp_i2c_update(&self) {
+        self.i2c.ctr().modify(|_, w| w.conf_upgate().set_bit());
+    }
+
+    fn reset_fifo(&self) {
+        self.i2c
+            .fifo_conf()
+            .modify(|_, w| w.tx_fifo_rst().set_bit());
+
+        self.i2c
+            .fifo_conf()
+            .modify(|_, w| w.tx_fifo_rst().clear_bit());
+
+        self.i2c
+            .fifo_conf()
+            .modify(|_, w| w.rx_fifo_rst().set_bit());
+
+        self.i2c
+            .fifo_conf()
+            .modify(|_, w| w.rx_fifo_rst().clear_bit());
+    }
+
+    fn wait_for_completion(&self) -> Result<(), Error> {
+        loop {
+            let interrupts = self.i2c.int_status().read();
```

Comment:
here and probably in other places: usually we use `int_raw` - then we don't need to enable / disable interrupts. is there a reason we cannot use `int_raw` here?

Reply:
I tried to change the interrupts from `status` to `raw` and it reads wrong data 🤷‍♂️